File: modules/native-geometry/android/src/main/java/expo/modules/nativegeometry/NativeGeometryModule.kt
package expo.modules.nativegeometry

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Production Expo Module — a thin delegator over [GeosBridge], which owns the
 * GEOS JNI layer. Keeping the GEOS logic in [GeosBridge] lets the instrumented
 * (`androidTest`) suite exercise the real binary without the RN bridge.
 */
class NativeGeometryModule : Module() {

    override fun definition() = ModuleDefinition {
        Name("NativeGeometry")

        Function("geosVersion") {
            GeosBridge.version()
        }

        // -- ABI version handshake (G5 follow-up) ------------------------------
        Function("nativeAbiVersion") { GeosBridge.NATIVE_ABI_VERSION }

        Function("bufferWKB") { wkb: ByteArray, distance: Double, quadrantSegments: Int ->
            GeosBridge.buffer(wkb, distance, quadrantSegments)
        }

        // -- Overlay ops: binary --------------------------------------------
        Function("differenceWKB") { wkbA: ByteArray, wkbB: ByteArray ->
            GeosBridge.difference(wkbA, wkbB)
        }

        Function("unionWKB") { wkbA: ByteArray, wkbB: ByteArray ->
            GeosBridge.union(wkbA, wkbB)
        }

        Function("intersectionWKB") { wkbA: ByteArray, wkbB: ByteArray ->
            GeosBridge.intersection(wkbA, wkbB)
        }

        // -- Overlay op: unary union ----------------------------------------
        Function("unaryUnionWKB") { wkb: ByteArray ->
            GeosBridge.unaryUnion(wkb)
        }
    }
}

File: modules/native-geometry/ios/NativeGeometryModule.swift
import ExpoModulesCore

public class NativeGeometryModule: Module {
    public func definition() -> ModuleDefinition {
        Name("NativeGeometry")

        Function("geosVersion") { () -> String in
            return GeosCore.version()
        }

        // Bump whenever the WKB function surface changes (e.g. new overlay ops).
        // JS side has EXPECTED_NATIVE_ABI; mismatch → loud rebuild warning.
        Function("nativeAbiVersion") { () -> Int in
            return 2 // G5 overlay ops
        }

        Function("bufferWKB") {
            (wkb: Data, distance: Double, quadrantSegments: Int) -> Data? in
            return GeosCore.buffer(
                wkb: wkb, distance: distance, quadrantSegments: quadrantSegments)
        }

        Function("differenceWKB") {
            (wkbA: Data, wkbB: Data) -> Data? in
            return GeosCore.difference(wkbA: wkbA, wkbB: wkbB)
        }

        Function("unionWKB") {
            (wkbA: Data, wkbB: Data) -> Data? in
            return GeosCore.union(wkbA: wkbA, wkbB: wkbB)
        }

        Function("intersectionWKB") {
            (wkbA: Data, wkbB: Data) -> Data? in
            return GeosCore.intersection(wkbA: wkbA, wkbB: wkbB)
        }

        Function("unaryUnionWKB") {
            (wkb: Data) -> Data? in
            return GeosCore.unaryUnion(wkb: wkb)
        }
    }
}
